components/push/src/subscriber.rs

```diff
@@ -161,6 +163,33 @@ impl PushManager {
             .map_err(|e| ErrorKind::TranscodingError(format!("{:?}", e)).into())
     }
 
+    /// Fetch new endpoints for a list of channels.
+    pub fn regenerate_endpoints(&mut self) -> error::Result<Vec<DispatchInfo>> {
+        if self.conn.uaid.is_none() {
+            return Err(ErrorKind::GeneralError("No subscriptions defined yet.".into()).into());
+        }
+        let uaid = self.conn.uaid.clone().unwrap();
+        let channels = self.store.get_channel_list(&uaid)?;
+        let mut results: Vec<DispatchInfo> = Vec::new();
+        for channel in channels {
+            if let Some(record) = self.store.get_record_by_chid(&channel)? {
```

Comment:
For the record: make sure we are Ok with this call failing or the ones below fails the whole method.

Reply:
Thanks for calling that out. This is something I needed @jrconlin 's comment on.

Reply:
To be honest, I'm wondering if auto-regenerating the subscriptions may not be a good idea. The reason being that the associated endpoint may change (as could the encryption key). The WebPush protocol normally requires that the WPA get a [`pushsubscriptionchange`](https://developer.mozilla.org/en-US/docs/Web/API/ServiceWorkerGlobalScope/pushsubscriptionchange_event) event, indicating that it needs to re-establish a given subscription, including sending the new subscription info block back to it's server. 

What that would mean in this case would be going through the PWA list (so each app associated with a ChannelID and sending the `pushsubscriptionchange` event. It'd be less clever, but more standardized.

Reply:
> What that would mean in this case would be going through the PWA list (so each app associated with a ChannelID and sending the pushsubscriptionchange event. It'd be less clever, but more standardized.

We no longer maintain a list of chids since we went from a 1 to n number of subscriptions. On the a-c side, I imagined invoking the `ushsubscriptionchange` event for each chid in the list that came back.

I'm not sure I know how to do it otherwise.

Reply:
> I'm wondering if auto-regenerating the subscriptions may not be a good idea. 

How I think I understand better of what you mean. What we really want is just the chid/scope of the subscription for notifying the apps. Maybe @eoger was right and I should have just created a new construct for just those instead of re-using the `DispatchInfo` type.

Reply:
Perhaps. I suppose part of it is that there's kind of a "knowledge" leak that needs to happen between the internal bits that do the encryption, and the javascript bits that do the actual consumption. At one point, I think the design was that the push component would help in that, but things stopped when there wasn't a clean way to do that, and that might have been the bit that was DispatchInfo. 

The nice thing is that i will not be sad about anything being taken behind the woodshed, so feel free to change-up whatever you need.

Reply:
> make sure we are Ok with this call failing or the ones below fails the whole method.

@jrconlin RE this issue, what do you think we should do if there is a failure in calling `get_record_by_chid`? I guess we shouldn't fail the entire call stack in the method?

Reply:
I'm not so sure. If that function fails, it means that the local data store is either corrupted or invalid, since the data associated with that channelID could not be found. At that point, it's probably worth forcing the refresh so that the push storage DB can be rebuilt.